### PR TITLE
chore: hash images

### PR DESCRIPTION
**Pull request by molant**
Adds an extra step to the precompile step to hash all the images so we can use the CDN and change the max-age values in another PR

### CODE REVIEW

**Comment by molant** [comment]
The images from the App section are not hashed, I can do that as well as part of the precompile step

File: script/precompile-assets.js [hunk 1]
```diff
@@ -62,6 +75,53 @@ function precompileCss() {
...
+  })
+
+  const images = [...appImages, ...websiteImages]
+
+  const manifest = {}
+
+  for (const image of images) {
+    const basename = path.basename(image)
+    const extension = path.extname(image)
+    const filename = basename.replace(extension, '')
```
Comment (MarshallOfSound) [inline script/precompile-assets.js:94]:
```suggestion
    const filename = path.basename(image, path.extname(image))
```

File: script/precompile-assets.js [hunk 2]
```diff
@@ -62,6 +75,53 @@ function precompileCss() {
   })
 }
 
+async function precompileImages() {
+  const websiteImages = await globby(`${PATHS.imagesDir}/**/*`, {
```
Comment (MarshallOfSound) [inline script/precompile-assets.js:79]:
```suggestion
  const websiteImages = await globby(path.join(PATHS.imagesDir, '**', '*'), {
```

We can support windows here 😄

File: script/precompile-assets.js [hunk 3]
```diff
@@ -62,6 +75,53 @@ function precompileCss() {
...
+    onlyFiles: true,
+  })
+
+  const appImages = await globby(`${PATHS.appImgDir}/**/*.png`, {
```
Comment (MarshallOfSound) [inline script/precompile-assets.js:83]:
```suggestion
  const appImages = await globby(path.join(PATHS.appImgDir, '**', '*.png'), {
```

File: script/precompile-assets.js [hunk 4]
```diff
@@ -62,6 +75,53 @@ function precompileCss() {
...
+    const basename = path.basename(image)
+    const extension = path.extname(image)
+    const filename = basename.replace(extension, '')
+    const content = await fs.readFile(image)
+    // We could optimize the images here and save a few bytes
+    const hash = calculateHash(content)
+    const imageDestination = image
+      .replace(PATHS.imagesDir, PATHS.imageDestinationDir)
+      .replace(PATHS.appImgDir, PATHS.appImgDestinationDir)
+      .replace(basename, `${filename}.${hash}${extension}`)
```
Comment (MarshallOfSound) [inline script/precompile-assets.js:101]:
```suggestion
    const imageDestination = path.resolve(
      path.dirname(image
        .replace(PATHS.imagesDir, PATHS.imageDestinationDir)
        .replace(PATHS.appImgDir, PATHS.appImgDestinationDir)
      ),
      `${filename}.${hash}${extension}`
    )
```

**Review by MarshallOfSound — COMMENTED** [review]
Looking good

File: script/precompile-assets.js (hunk 2 above)
Comment (molant) [reply script/precompile-assets.js:79]: WSL has spoiled me 😅

**Comment by molant** [comment]
Thanks @MarshallOfSound !

**Comment by molant** [comment]
Actually on Windows the images still don't get hashed with those changes. Going to look into it.

**Comment by molant** [comment]
@MarshallOfSound I had to add a few more edits to make it work entirely on Windows:
* globby only supports `/` so I ended up using `path.posix.join`
* the final `manifest.json` ended up with a mix of `\\` and `/` and the incorrect path if using resolve so ended up using the previous code

**Review by MarshallOfSound — APPROVED** [review]

File: server.js [hunk 5]
```diff
@@ -90,23 +90,42 @@ if (process.env.NODE_ENV === 'production') {
...
+  const imagesManifest = require(path.join(
+    __dirname,
+    'precompiled',
+    'images',
+    'manifest.json'
+  ))
+  hbs.registerHelper('static-asset', (type, ...parts) => {
+    // `parts` should be at minimum [name, function]
+    // but it could also be [part1, part2, part3, function ]
+    // if we want to link to dynamic images
```
Comment (erickzhao) [inline server.js:102]: What does `function` mean in this context? Is it the file extension?

**Review by erickzhao — APPROVED** [review]

File: server.js (hunk 5 above)
Comment (molant) [reply server.js:102]: No, it's actually a function :)

File: scripts/search.js [hunk 6]
```diff
@@ -17,10 +17,6 @@ const types = [
     name: 'package',
     path: '/community',
   },
-  {
```
[the remark this answers is not in the document]
Comment (molant) [reply scripts/search.js:20]: Personally I prefer not to have any commented code and leave as much information in the commit message for future reference but whatever you prefer :)

**Pull request merged**